File: euxis-core/src/euxis_core/mesh/router.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2024-2026 Sebastien Rousseau ᛫ https://sebastienrousseau.com

"""FinOps Router for cost-aware model provider selection."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

LOGGER = logging.getLogger("euxis.core.router")

@dataclass
class ProviderMetrics:
    name: str
    cost_per_1k_tokens: float # in USD
    avg_latency_ms: int
    reliability_score: float # 0.0 to 1.0

class FinOpsRouter:
    """Intelligent router that balances cost, speed, and reliability."""

    def __init__(self, budget_limit: float = 10.0):
        self.budget_limit = budget_limit
        self.current_spend = 0.0
        # Default 2026 model pricing / performance stats
        self.providers = [
            ProviderMetrics("ollama", 0.000, 150, 0.95),    # Local Edge AI
            ProviderMetrics("groq", 0.0001, 50, 0.98),      # Fast LPU
            ProviderMetrics("anthropic", 0.015, 800, 0.99), # High reasoning
            ProviderMetrics("openai", 0.010, 600, 0.99),    # Balanced
        ]

    def select_provider(self, task_complexity: str, priority: str = "balanced") -> str:
        """Select the best provider for a given task and priority."""
        LOGGER.info(f"Routing task (complexity: {task_complexity}, priority: {priority})")
        
        if task_complexity == "low":
            # Always prefer local for low complexity
            return "ollama"
            
        if priority == "speed":
            # Sort by latency
            sorted_providers = sorted(self.providers, key=lambda p: p.avg_latency_ms)
            return sorted_providers[0].name
            
        if priority == "cost":
            # Sort by cost
            sorted_providers = sorted(self.providers, key=lambda p: p.cost_per_1k_tokens)
            return sorted_providers[0].name
            
        # Default: Weighted score balancing reliability, cost, and latency.
        best_provider = "ollama"
        best_score = -100.0
        
        for p in self.providers:
            # Normalized values for scoring
            cost_factor = p.cost_per_1k_tokens * 100 # Scale up for significance
            latency_factor = p.avg_latency_ms / 1000
            
            score = (p.reliability_score * 10.0) - (cost_factor * 5.0) - (latency_factor * 2.0)
            
            if score > best_score:
                best_score = score
                best_provider = p.name
                
        LOGGER.info(f"FinOps Router selected: {best_provider} (score: {best_score:.2f})")
        return best_provider

    def track_usage(self, provider_name: str, tokens: int):
        """Update spend tracking."""
        provider = next((p for p in self.providers if p.name == provider_name), None)
        if provider:
            cost = (tokens / 1000) * provider.cost_per_1k_tokens
            self.current_spend += cost
            LOGGER.info(f"Current session spend: ${self.current_spend:.4f}")

File: euxis-cpp/euxis-core-cpp/src/router.cpp
#include "euxis/core/router.hpp"

#include <algorithm>
#include <string>
#include <vector>

#include <spdlog/spdlog.h>

namespace euxis::core {

FinOpsRouter::FinOpsRouter(double budget_limit) : budget_limit_(budget_limit) {
    providers_ = {
        {"ollama", 0.000, 150, 0.95},
        {"groq", 0.0001, 50, 0.98},
        {"anthropic", 0.015, 800, 0.99},
        {"openai", 0.010, 600, 0.99},
    };
}

auto FinOpsRouter::select_provider(const std::string& task_complexity,
                                   const std::string& priority) -> std::string {
    spdlog::info("Routing task (complexity: {}, priority: {})",
                 task_complexity, priority);

    if (task_complexity == "low") {
        return "ollama";
    }

    if (priority == "speed") {
        auto it = std::ranges::min_element(
            providers_, {}, &ProviderMetrics::avg_latency_ms);
        return it->name;
    }

    if (priority == "cost") {
        auto it = std::ranges::min_element(
            providers_, {}, &ProviderMetrics::cost_per_1k_tokens);
        return it->name;
    }

    // Weighted score balancing reliability, cost, and latency
    std::string best = "ollama";
    double best_score = -100.0;
    for (const auto& p : providers_) {
        const double cost_factor = p.cost_per_1k_tokens * 100.0;
        const double latency_factor = static_cast<double>(p.avg_latency_ms) / 1000.0;
        const double score = (p.reliability_score * 10.0) -
                       (cost_factor * 5.0) - (latency_factor * 2.0);
        if (score > best_score) {
            best_score = score;
            best = p.name;
        }
    }

    spdlog::info("FinOps Router selected: {} (score: {:.2f})", best, best_score);
    return best;
}

void FinOpsRouter::track_usage(const std::string& provider_name, int tokens) {
    for (const auto& p : providers_) {
        if (p.name == provider_name) {
            const double cost =
                (static_cast<double>(tokens) / 1000.0) * p.cost_per_1k_tokens;
            current_spend_ += cost;
            spdlog::info("Current session spend: ${:.4f}", current_spend_);
            return;
        }
    }
}

void FinOpsRouter::track_session_usage(const std::string& session_id,
                                        const std::string& agent_id,
                                        const std::string& model,
                                        int input_tokens, int output_tokens) {
    const int total = input_tokens + output_tokens;
    double cost = 0.0;
    // Estimate cost based on known providers
    for (const auto& p : providers_) {
        if (model.find(p.name) != std::string::npos) {
            cost = (static_cast<double>(total) / 1000.0) * p.cost_per_1k_tokens;
            break;
        }
    }
    if (cost == 0.0) {
        // Default estimate
        cost = static_cast<double>(total) / 1000.0 * 0.001;
    }

    session_usage_[session_id].push_back({
        .agent_id = agent_id,
        .model = model,
        .input_tokens = input_tokens,
        .output_tokens = output_tokens,
        .cost = cost,
    });
    current_spend_ += cost;
}

auto FinOpsRouter::session_cost(const std::string& session_id) const -> double {
    const auto it = session_usage_.find(session_id);
    if (it == session_usage_.end()) {
        return 0.0;
    }

    double total = 0.0;
    for (const auto& r : it->second) {
        total += r.cost;
    }
    return total;
}

auto FinOpsRouter::check_budget(const std::string& session_id,
                                 double limit) const -> bool {
    return session_cost(session_id) <= limit;
}

} // namespace euxis::core
